src/Illuminate/Routing/RouteDependencyResolverTrait.php

```diff
@@ -43,50 +43,111 @@ protected function resolveClassMethodDependencies(array $parameters, $instance,
             }
+
+            $arguments[] = $instance;
         }
 
-        return $parameters;
+        // Remove all found dependencies from $parameters array,
+        // this way we can inject all non object parameters in their respective order
+        $parameters = $this->cleanDependenciesFromParameters($parameters, $classNames);
+
+        if (! empty($parameters)) {
+            return $this->mergeRemainingParameters($arguments, $parameters);
+        }
+
+        return $arguments;
     }
 
     /**
      * Attempt to transform the given parameter into a class instance.
      *
      * @param  \ReflectionParameter  $parameter
      * @param  array  $parameters
-     * @param  array  $originalParameters
      * @return mixed
      */
-    protected function transformDependency(ReflectionParameter $parameter, $parameters, $originalParameters)
+    protected function transformDependency(ReflectionParameter $parameter, array $parameters)
     {
-        $class = $parameter->getClass();
+        // If the current parameter shouldn't be a class this method will return null
+        if ($class = $parameter->getClass()) {
+
+            // If the parameter has a type-hinted class, we will check to see if it is already in
+            // the list of parameters. If it is we return that class; otherwise, we resolve it here.
+            if ($this->alreadyInParameters($class->name, $parameters)) {
+                return $this->fetchParameterByClassName($class->name, $parameters);
+            }
 
-        // If the parameter has a type-hinted class, we will check to see if it is already in
-        // the list of parameters. If it is we will just skip it as it is probably a model
-        // binding and we do not want to mess with those; otherwise, we resolve it here.
-        if ($class && ! $this->alreadyInParameters($class->name, $parameters)) {
             return $this->container->make($class->name);
         }
     }
 
     /**
+     * Remove known classes from an array of parameters.
+     *
+     * @param  array $originalParameters
+     * @param  array $classNames
+     * @return array
+     */
+    protected function cleanDependenciesFromParameters(array $originalParameters, array $classNames)
+    {
+        $parameters = [];
+        // Remove any duplicates from class list
+        $classNames = array_unique($classNames);
+
+        // Create a new parameters array leaving out any classes from the class list
+        foreach ($originalParameters as $key => $parameter) {
+            if (! is_object($parameter) || ! in_array(get_class($parameter), $classNames)) {
+                $parameters[$key] = $parameter;
+            }
+        }
+
+        return $parameters;
+    }
+
+    /**
+     * Merge route parameters into method arguments.
+     *
+     * @param  array $arguments
+     * @param  array $parameters
+     * @return array
+     */
+    protected function mergeRemainingParameters($arguments, $parameters)
```

Comment:
Here we can add type hinting: `protected function mergeRemainingParameters(array $arguments, array $parameters)`